fuse_models/include/fuse_models/common/sensor_proc.hpp

```diff
@@ -1225,6 +2097,117 @@ inline bool processAccelWithCovariance(
+ * given \p transaction. The acceleration data is extracted from the \p acceleration message.
+ * The data will be automatically transformed into the \p target_frame before it is used.
+ *
+ * @param[in] source - The name of the sensor or motion model that generated this constraint
+ * @param[in] device_id - The UUID of the machine
+ * @param[in] acceleration - The AccelWithCovarianceStamped message from which we will extract the
+ *                           acceleration data
+ * @param[in] loss - The loss function for the 3D linear acceleration constraint generated
+ * @param[in] indices - The indices of the linear acceleration vector to use. If empty, no
+ *                     linear acceleration constraint is added
+ * @param[in] target_frame - The frame ID into which the acceleration data will be transformed
+ *                           before it is used
+ * @param[in] tf_buffer - The transform buffer with which we will lookup the required transform
+ * @param[in] validate - Whether to validate the measurements or not. If the validation fails no
+ *                       constraint is added
+ * @param[out] transaction - The generated variables and constraints are added to this transaction
+ * @return true if any constraints were added, false otherwise
+ */
+inline bool processAccel3DWithCovariance(
+  const std::string & source,
+  const fuse_core::UUID & device_id,
+  const geometry_msgs::msg::AccelWithCovarianceStamped & acceleration,
+  const fuse_core::Loss::SharedPtr & loss,
+  const std::string & target_frame,
+  const std::vector<size_t> & indices,
+  const tf2_ros::Buffer & tf_buffer,
+  const bool validate,
+  fuse_core::Transaction & transaction,
+  const rclcpp::Duration & tf_timeout = rclcpp::Duration(0, 0))
+{
+  // Make sure we actually have work to do
+  if (indices.empty()) {
+    return false;
+  }
+
+  geometry_msgs::msg::AccelWithCovarianceStamped transformed_message;
+  if (target_frame.empty()) {
+    transformed_message = acceleration;
+  } else {
+    transformed_message.header.frame_id = target_frame;
+
+    if (!transformMessage(tf_buffer, acceleration, transformed_message, tf_timeout)) {
+      RCLCPP_WARN_STREAM_SKIPFIRST_THROTTLE(
+        rclcpp::get_logger("fuse"), sensor_proc_clock, 10.0,
+        "Failed to transform acceleration message with stamp " <<
+          rclcpp::Time(acceleration.header.stamp).nanoseconds()
+                                                               << ". Cannot create constraint.");
+      return false;
+    }
+  }
+
+  // Create the full mean vector and covariance for the constraint
+  fuse_core::Vector3d accel_mean;
+  accel_mean << 
+    transformed_message.accel.accel.linear.x, 
+    transformed_message.accel.accel.linear.y,
+    transformed_message.accel.accel.linear.z;
+
+  Eigen::Map<const fuse_core::Matrix3d> accel_covariance_map(transformed_message.accel.covariance.data());
```

Comment:
```suggestion
  Eigen::Map<const fuse_core::Matrix3d, 0, Eigen::Stride<6, 1>> accel_covariance_map(transformed_message.accel.covariance.data());
```

I think you need this stride here. Without it, feeding it a cov mat with 1,2,3,4,5,6,7,8,9 resulted in:
```
1 2 3
0 0 0
4 5 6
```

With it, it results in 
```
1 2 3
4 5 6
7 8 9
```
As you'd expect. This is because without the stride term, Eigen has no way to know you want it to treat the double[36] as a 6x6, so it just grabs the first row (1 2 3 0 0 0) and then the second row (4 5 6).

Reply:
You can verify this by just launching an IMU node and then manually publishing to its topic. I found it in the process of adding the tutorials I mentioned.

Reply:
I agree! thanks